build: The vcpkg tool has introduced a proper way to use manifests

sipsorcery opened the pull request:
The vcpkg tool has introduced a proper way to use [manifests](https://devblogs.microsoft.com/cppblog/vcpkg-accelerate-your-team-development-environment-with-binary-caching-and-manifests/). This PR replaces the custom text file mechanism with the new manifest approach.

It is planned that vckpg manifests will include the ability to version dependencies in the future. Dependency versions would solve a number of issues that currently require workarounds with the appveyor CI.

fanquake commented:
Concept ACK

hebasto commented:
Concept ACK.

MarcoFalke commented on build_msvc/vcpkg.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "bitcoin-core",
+  "version-string": "0.20.99",
```
Does this need to be bumped at the same time the other version string is bumped?

MarcoFalke replied:
"End users can set these fields to whatever they want, but the name must be all lowercase characters." From the Link in OP


So I'd suggest to set this to something not entangled with the bitcoin core version. Maybe just `"1"`?

sipsorcery replied:
You beat me to it and as you've mentioned it's arbitrary. The version field (https://vcpkg.readthedocs.io/en/latest/specifications/manifests/) seems to be intended for use when creating vcpkg port files rather than consuming them. It is a mandatory field though so has to be set to something. I've gone ahead and set it to `1` to avoid any perception it relates to any other release or version numbers.

MarcoFalke commented:
<!--e57a25ab6845829454e8d69fc972939a-->

The following sections might be updated with supplementary metadata relevant to reviewers and maintainers.

<!--174a7506f384e20aa4161008e828411d-->
### Conflicts
Reviewers, this pull request conflicts with the following ones:

* #19077 (wallet: Add sqlite as an alternative wallet database and use it for new descriptor wallets by achow101)

If you consider this pull request important, please also help to review the conflicting pull requests. Ideally, start with the one that should be merged first.

fanquake commented on build_msvc/README.md:
```diff
@@ -12,10 +12,9 @@ Quick Start
-vcpkg install --triplet x64-windows-static berkeleydb boost-filesystem boost-multi-index boost-signals2 boost-test boost-thread libevent[thread] zeromq double-conversion
-vcpkg integrate install
+cd build_msvc
 py -3 build_msvc\msvc-autogen.py
```
```suggestion
py -3 msvc-autogen.py
```

MarcoFalke commented on .appveyor.yml:
```diff
@@ -11,26 +11,20 @@ environment:
-      Write-Host "vcpkg packages installed successfully."
-      .\vcpkg integrate install
-      cd "$env:APPVEYOR_BUILD_FOLDER"
+      cd $env:APPVEYOR_BUILD_FOLDER
```
Is it needed to remove the `"`?

sipsorcery replied:
No, reverted, thx.

maflcko reviewed (approved):
Please squash and format the commit message with subject line and body

fanquake commented on build_msvc/README.md:
```diff
@@ -28,14 +27,7 @@ Options for installing the dependencies in a Visual Studio compatible manner are
-- libevent
-- Qt5
-- ZeroMQ
+The [external dependencies](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) required for building are listed in the `build_msvc/vcpkg.json` file. The `msbuild` project files are configured to automatically install the `vcpg` dependencies.
```
```suggestion
The [external dependencies](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) required for building are listed in the `build_msvc/vcpkg.json` file. The `msbuild` project files are configured to automatically install the `vcpkg` dependencies.
```

sipsorcery replied:
Thx, fixed.

maflcko reviewed (approved):
Approach ACK b92a074570e06b07fdd87a8ed03d1cf7139c7407

fanquake commented:
This looks good now. However can you follow up with:

> [Please squash and format the commit message with subject line and body](https://github.com/bitcoin/bitcoin/pull/19960#pullrequestreview-492180598)

sipsorcery committed:
Update msvc build to use new vcpkg manifest

The vcpkg tool has introduced a proper way to use manifests, https://devblogs.microsoft.com/cppblog/vcpkg-accelerate-your-team-development-environment-with-binary-caching-and-manifests/. This PR replaces the custom text file mechanism with the new manifest approach.

It is planned that vckpg manifests will include the ability to version dependencies in the future. Dependency versions would solve a number of issues that currently require workarounds with the appveyor CI.

Set vcpkg manifest version to 1 to avoid any perception it's related to any release or other version numbering.

sipsorcery commented:
I've adjusted my commit message to have a better lead (subject) line. Was that the problem?

fanquake reviewed (approved):
ACK 712f95d3324d02310dd468e7bfd1e1b0df432e77 - This is a nice simplification. I tested this in a Windows VM; packages were downloaded and installed automatically as required:

![vcpkg](https://user-images.githubusercontent.com/863730/94562211-a23a9e80-0297-11eb-81f2-76d727363cab.png)

> I've adjusted my commit message to have a better lead (subject) line. Was that the problem?

Thanks. Yes it was.

One other thought. Is there any Appveyor related cache me might need to nuke to make sure this is really working there as expected?

hebasto commented on .appveyor.yml:
```diff
@@ -11,25 +11,19 @@ environment:
   QT_LOCAL_PATH: 'C:\Qt5.9.8_x64_static_vs2019'
   VCPKG_INSTALL_PATH: 'C:\tools\vcpkg\installed'
-  VCPKG_COMMIT_ID: 'f3f329a048eaff759c1992c458f2e12351486bc7'
+  VCPKG_COMMIT_ID: '13590753fec479c5b0a3d48dd553dde8d49615fc'
```
Why the latest [release](https://github.com/microsoft/vcpkg/releases), i.e., https://github.com/microsoft/vcpkg/commit/56fffbe49dfb4dd8fae0940f272c5fd2b86be991, is not used?

sipsorcery replied:
No good reason. I've always built from the vcpkg `master` branch, checked that the dependencies work and used that commit. The `vcpkg` repo is a bit different to code only repos in that commits are made for the library port configs as well as the source code use to build the `vcpkg` executable.

hebasto commented on build_msvc/.gitignore:
```diff
@@ -24,3 +24,4 @@ libtest_util/libtest_util.vcxproj
 */Win32
 libbitcoin_qt/QtGeneratedFiles/*
 test_bitcoin-qt/QtGeneratedFiles/*
+vcpkg_installed
```
nit: add EOL?

hebasto commented on build_msvc/README.md:
```diff
@@ -28,14 +27,7 @@ Options for installing the dependencies in a Visual Studio compatible manner are
-- libevent
-- Qt5
-- ZeroMQ
+The [external dependencies](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) required for building are listed in the `build_msvc/vcpkg.json` file. The `msbuild` project files are configured to automatically install the `vcpkg` dependencies.
```
```suggestion
The [external dependencies](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) required for building are listed in the [`build_msvc/vcpkg.json`](https://github.com/bitcoin/bitcoin/tree/master/build_msvc/vcpkg.json) file. The `msbuild` project files are configured to automatically install the `vcpkg` dependencies.
```

sipsorcery commented:
> One other thought. Is there any Appveyor related cache me might need to nuke to make sure this is really working there as expected?

Nope. All appveyor caching has already been removed. Predominantly due to nasty issues with the cached vcpkg dependencies getting messed up between different PR branches.

With this new manifest approach it might be feasible to re-introduce a cache of the vcpkg dependencies at some point, and save 20mins per build, but best to leave that until the manifest change is bedded down.

hebasto commented on build_msvc/vcpkg.json:
```diff
@@ -0,0 +1,19 @@
+    },
+    "zeromq"
+  ]
+}
```
Add EOL?

hebasto reviewed (commented):
Approach ACK 712f95d3324d02310dd468e7bfd1e1b0df432e77, I've verified that changes comply MS docs:
- https://github.com/microsoft/vcpkg/blob/master/docs/specifications/manifests.md
- https://devblogs.microsoft.com/cppblog/vcpkg-accelerate-your-team-development-environment-with-binary-caching-and-manifests

hebasto commented:
@sipsorcery Could you look into the recent "error MSB3073" AppVeyor fails?

sipsorcery commented:
So far haven't found anything likely that changed around 5 hours ago when the builds started failing. Will keep digging.

MarcoFalke commented:
This may have increased the average `msbuild` time (at least on appveyor)

sipsorcery commented:
@MarcoFalke yes I see what you mean. It seems that build takes approximately 10 minutes longer subsequent to the change in this PR.

One possibility is the `vcpkg` dependencies may now be being built for both debug and release. Previously the line below restricted them to only release. I'll do some testing.

````
Add-Content "C:\tools\vcpkg\triplets\$env:PLATFORM-windows-static.cmake" "set(VCPKG_BUILD_TYPE release)"
````

sipsorcery commented:
Finally got around to running the build tests to compare the build durations using the original vcpkg install mechanism and the new vcpkg manifest mechanism.

The original way seems to be a few minutes faster but not a smoking gun for a 10 minute increase. Perhaps the manifest + the removal of the `set(VCPKG_BUILD_TYPE release)` could account for it. The `VCPKG_BUILD_TYPE` setting should be put back and I'll create a PR for that.

**Build runs with vcpkg installed from command line:**
**Average duration approx. 25 minutes.**
```
Run 1:
c:\Tools\vcpkg>vcpkg install --triplet x64-windows-static berkeleydb boost-filesystem boost-multi-index boost-process boost-signals2 boost-test boost-thread libevent[thread] zeromq double-conversion sqlite3 -> 14:12
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 10:37

Run 2:
c:\Tools\vcpkg>vcpkg install --triplet x64-windows-static berkeleydb boost-filesystem boost-multi-index boost-process boost-signals2 boost-test boost-thread libevent[thread] zeromq double-conversion sqlite3 -> 16:10
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 09:40

Run 3:
c:\Tools\vcpkg>vcpkg install --triplet x64-windows-static berkeleydb boost-filesystem boost-multi-index boost-process boost-signals2 boost-test boost-thread libevent[thread] zeromq double-conversion sqlite3 -> 15:29
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 09:36

Run 4:
c:\Tools\vcpkg>vcpkg install --triplet x64-windows-static berkeleydb boost-filesystem boost-multi-index boost-process boost-signals2 boost-test boost-thread libevent[thread] zeromq double-conversion sqlite3 -> 15:12
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 09:39
```

**Build runs with vcpkg installed using manifest mechanism:**
**Average duration approx. 28 minutes.**
```
Run 1:
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 26:58

Run 2:
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 27:00

Run 3:
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 27:15

Run 4:
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 30:31

Run 5:
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 30:31

Run 6:
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msvc-autogen.py
c:\Dev\github\sipsorcery_bitcoin\build_msvc>msbuild /m bitcoin.sln /p:Configuration=Release /p:Platform=x64 /t:clean,build -> 24:51
```
